### PR TITLE
Issue2916499

### PR DESCRIPTION
**Pull request by petrillek**
Linear Gradient and Custom linear gradient modifiers are ready. 
Please review @kybersoft. :)

### CODE REVIEW

File: modules/modifiers_custom_linear_gradient/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Custom Linear gradient modifier
+
+## Overview
+This module implements a Modifier which allows you to use colors to set a 
+linear gradient. You can set multiple colors, they will be evenly distributed
+ along a customizable direction. If only one color is provided it will be set
```
Comment (kybersoft) [inline modules/modifiers_custom_linear_gradient/README.md]: These empty spaces at the line start (2x) are not needed (also inside modifiers_linear_gradient), please remove them.

File: modules/modifiers_custom_linear_gradient/modifiers_custom_linear_gradient.info.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+name: Custom Linear Gradient Modifier
+type: module
+description: Provides a Modifier to set the background gradient on an element using custom colors.
+core: 8.x
+package: Modifiers
+dependencies:
+  - modifiers
```
Comment (kybersoft) [inline modules/modifiers_custom_linear_gradient/modifiers_custom_linear_gradient.info.yml]: Dependencies need to have prefix, you can copy them from modifiers_custom_colors.info.yml.

File: modules/modifiers_linear_gradient/modifiers_linear_gradient.info.yml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+name: Linear Gradient Modifier
+type: module
+description: Provides a Modifier to set the background gradient on an element using colors from library.
+core: 8.x
+package: Modifiers
+dependencies:
+  - modifiers
```
Comment (kybersoft) [inline modules/modifiers_linear_gradient/modifiers_linear_gradient.info.yml]: Dependencies need to have prefix, you can copy them from modifiers_colors.info.yml.

File: modules/modifiers_custom_linear_gradient/modifiers_custom_linear_gradient.info.yml [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  - paragraphs
+  - taxonomy
```
Comment (kybersoft) [inline modules/modifiers_custom_linear_gradient/modifiers_custom_linear_gradient.info.yml]: Taxonomy is not dependency of modifiers_custom_linear_gradient module, only modifiers_linear_gradient.

File: modules/modifiers_custom_linear_gradient/src/Plugin/modifiers/CustomLinearGradientModifier.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
+<?php
+
+namespace Drupal\modifiers_custom_linear_gradient\Plugin\modifiers;
+
+use Drupal\modifiers\Modification;
+use Drupal\modifiers\ModifierPluginBase;
+
+/**
+ * Provides a Modifier to set the linear gradient on an element using custom colors.
```
Comment (kybersoft) [inline modules/modifiers_custom_linear_gradient/src/Plugin/modifiers/CustomLinearGradientModifier.php]: This line exceeds 80 characters, please shorten that - phpcs need short description on single line. I would stay with "Provides a Modifier to set the linear gradient on an element."

File: modules/modifiers_linear_gradient/src/Plugin/modifiers/LinearGradientModifier.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
+<?php
+
+namespace Drupal\modifiers_linear_gradient\Plugin\modifiers;
+
+use Drupal\modifiers\Modification;
+use Drupal\modifiers\ModifierPluginBase;
+
+/**
+ * Provides a Modifier to set the linear gradient on an element using custom colors.
```
Comment (kybersoft) [inline modules/modifiers_linear_gradient/src/Plugin/modifiers/LinearGradientModifier.php]: This line exceeds 80 characters, please shorten that - phpcs need short description on single line. I would stay with "Provides a Modifier to set the linear gradient on an element."

File: modules/modifiers_linear_gradient/src/Plugin/modifiers/LinearGradientModifier.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+   * {@inheritdoc}
+   */
+  public static function modification($selector, array $config) {
+
+    $css = [];
+    $attributes = [];
+    $media = parent::getMediaQuery($config);
+
+    if (!empty($config['l_gradient_colors'])) {
+      $attributes['class'][] = 'modifiers-has-background';
```
Comment (kybersoft) [inline modules/modifiers_linear_gradient/src/Plugin/modifiers/LinearGradientModifier.php:29]: Please insert `$direction = '';` and one empty line right after this line.

File: modules/modifiers_linear_gradient/src/Plugin/modifiers/LinearGradientModifier.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+      if (!empty($config['l_gradient_direction'])) {
+        $direction = $config['l_gradient_direction'] . 'deg,';
+      }
+      // If there is only one color specified we replicate it in order to have
+      // one color fill.
+      if (count($config['l_gradient_colors']) == 1) {
+        $css[$media][$selector][] = 'background:' . $config['l_gradient_colors'][0];
+      }
+      else {
+        $css[$media][$selector][] = 'background:linear-gradient(' . $direction . implode(',', $config['l_gradient_colors']) . ')';
```
Comment (kybersoft) [inline modules/modifiers_linear_gradient/src/Plugin/modifiers/LinearGradientModifier.php]: Insert line break right after `'background:linear-gradient('`.

File: modules/modifiers_linear_gradient/src/Plugin/modifiers/LinearGradientModifier.php [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    $media = parent::getMediaQuery($config);
+
+    if (!empty($config['l_gradient_colors'])) {
+      $attributes['class'][] = 'modifiers-has-background';
+      if (!empty($config['l_gradient_direction'])) {
+        $direction = $config['l_gradient_direction'] . 'deg,';
+      }
+      // If there is only one color specified we replicate it in order to have
+      // one color fill.
+      if (count($config['l_gradient_colors']) == 1) {
```
Comment (kybersoft) [inline modules/modifiers_linear_gradient/src/Plugin/modifiers/LinearGradientModifier.php]: Please use "===".

File: modules/modifiers_custom_linear_gradient/src/Plugin/modifiers/CustomLinearGradientModifier.php [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+   * {@inheritdoc}
+   */
+  public static function modification($selector, array $config) {
+
+    $css = [];
+    $attributes = [];
+    $media = parent::getMediaQuery($config);
+
+    if (!empty($config['cl_gradient_colors'])) {
+      $attributes['class'][] = 'modifiers-has-background';
```
Comment (kybersoft) [inline modules/modifiers_custom_linear_gradient/src/Plugin/modifiers/CustomLinearGradientModifier.php:29]: Please insert `$direction = '';` and one empty line right after this line.

File: modules/modifiers_custom_linear_gradient/src/Plugin/modifiers/CustomLinearGradientModifier.php [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+      if (!empty($config['cl_gradient_direction'])) {
+        $direction = $config['cl_gradient_direction'] . 'deg,';
+      }
+      // If there is only one color specified we replicate it in order to have
+      // one color fill.
+      if (count($config['cl_gradient_colors']) == 1) {
```
Comment (kybersoft) [inline modules/modifiers_custom_linear_gradient/src/Plugin/modifiers/CustomLinearGradientModifier.php]: Please use "===".

File: modules/modifiers_custom_linear_gradient/src/Plugin/modifiers/CustomLinearGradientModifier.php [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        $css[$media][$selector][] = 'background:' . $config['cl_gradient_colors'][0];
+      }
+      else {
+        $css[$media][$selector][] = 'background:linear-gradient(' . $direction . implode(',', $config['cl_gradient_colors']) . ')';
```
Comment (kybersoft) [inline modules/modifiers_custom_linear_gradient/src/Plugin/modifiers/CustomLinearGradientModifier.php]: Insert line break right after `'background:linear-gradient('`.

**Review by kybersoft — CHANGES_REQUESTED** [review]
Thank you Petr for great work! Please process these small  fixes.

**Comment by petrillek** [comment]
Thanks for the review @kybersoft . 
Changes have been done.

**Review by kybersoft — APPROVED** [review]

**Comment by kybersoft** [comment]
That looks much better, thank you @petrillek.
I checked also config files and there are some weird dependencies (`- ds` and `- workbench_moderation`) - can we get rid of them?

**Comment by kybersoft** [comment]
These 3 files should be reexported without "ds" and "workbench_moderation" modules:
- core.entity_view_display.paragraph.custom_linear_gradient_modifier.default.yml
- core.entity_form_display.paragraph.custom_linear_gradient_modifier.default.yml
- core.entity_form_display.paragraph.linear_gradient_modifier.default.yml

**Comment by petrillek** [comment]
Another itteration for you. ;) @kybersoft

**Comment by kybersoft** [comment]
Tested and everything is working correctly, thanks. Going to merge :+1:

**Pull request merged**